crates/sui-network/src/state_sync/mod.rs

```diff
@@ -832,42 +855,24 @@ where
         .buffered(checkpoint_header_download_concurrency);
 
     while let Some((maybe_checkpoint, next)) = request_stream.next().await {
+        debug_assert!(current.sequence_number().saturating_add(1) == next);
+
         // Verify the checkpoint
         let checkpoint = {
             let checkpoint = maybe_checkpoint
                 .ok_or_else(|| anyhow::anyhow!("no peers were able to help sync"))?;
-
-            if checkpoint.sequence_number() != next
-                || Some(current.digest()) != checkpoint.previous_digest()
-            {
-                return Err(anyhow::anyhow!("detected fork"));
-            }
-
-            let current_epoch = current.epoch();
-            if checkpoint.epoch() != current_epoch
-                && checkpoint.epoch() != current_epoch.saturating_add(1)
-            {
-                return Err(anyhow::anyhow!(
-                    "cannot verify checkpoint with too high of an epoch {}, current epoch {}",
-                    checkpoint.epoch(),
-                    current_epoch,
-                ));
-            }
-
-            if checkpoint.epoch() == current_epoch.saturating_add(1)
-                && current.next_epoch_committee().is_none()
-            {
-                return Err(anyhow::anyhow!(
-                    "next checkpoint claims to be from the next epoch but the latest verified \
-                    checkpoint does not indicate that it is the last checkpoint of an epoch"
-                ));
+            match verify_checkpoint(&current, &store, checkpoint) {
+                Ok(verified_checkpoint) => verified_checkpoint,
+                Err(checkpoint) => {
+                    // Remove the checkpoint from our temporary store so that we can try querying
+                    // another peer for a different one
```

Comment:
we should also blocklist this peer (at least for a period of time) as it could keep sending us invalid data. Doesn't need to be in this PR though

Reply:
We probably need to do more in this space, but for now i've changed this to mark the peer as not on the same chain as us, hopefully preventing us from paying any attention to them.